build-system/tasks/performance/download-images.js

```diff
@@ -0,0 +1,94 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const fs = require('fs');
+const {CONTROL, maybeCopyImageToCache, urlToCachePath} = require('./helpers');
+const {JSDOM} = require('jsdom');
+
+/**
+ * Lookup URL from cache. Look at tags that could use images. Download images
+ * with a local src to cache.
+ *
+ * @param {string} url
+ */
+function downloadImagesFromTags(url) {
+  const cachePath = urlToCachePath(url, CONTROL);
+  const document = fs.readFileSync(cachePath);
+  const dom = new JSDOM(document);
+
+  downloadImagesFromAmpImg(url, dom);
+  downloadImagesFromAmpVideo(url, dom);
+}
+
+/**
+ * Download locally stored images found in the amp-img tags
+ * from src and srcset
+ *
+ * @param {url} url
+ * @param {JSDOM} dom
+ */
+function downloadImagesFromAmpImg(url, dom) {
+  const imageTags = Array.from(dom.window.document.querySelectorAll('amp-img'));
+
+  for (const imageTag of imageTags) {
+    const src = imageTag.getAttribute('src');
+    const srcset = imageTag.getAttribute('srcset');
+    if (src) {
+      // Remove `?...` that may be used at the end of the src for uniqueness
+      const srcPath = src.split('?')[0];
+      maybeCopyImageToCache(url, srcPath);
+    }
+    if (srcset) {
+      // Get each source path, strip out responsive (1x, 2x..) found at the end
+      const sources = srcset.split(',').map((s) => s.trim().split(' ')[0]);
+
+      for (const path of sources) {
+        const srcPath = path.split('?')[0];
```

Comment:
super nit: since this is repeated, it may make more sense as part of `maybeCopyImageToCache`